File: proton-c/bindings/ruby/lib/reactor/connector.rb
#--
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#++

module Qpid::Proton::Reactor

  class Connector < Qpid::Proton::BaseHandler

    def initialize(connection, url, opts)
      @connection, @opts = connection, opts
      @urls = URLs.new(url) if url
      opts.each do |k,v|
        case k
        when :url, :urls, :address
          @urls = URLs.new(v) unless @urls
        when :reconnect
          @reconnect = v
        end
      end
      raise ::ArgumentError.new("no url for connect") unless @urls

      # TODO aconway 2017-08-17: review reconnect configuration and defaults
      @reconnect = Backoff.new() unless @reconnect
      @ssl_domain = SessionPerConnection.new # TODO seems this should be configurable
      @connection.overrides = self
      @connection.open
    end

    def on_connection_local_open(event)
      self.connect(event.connection)
    end

    def on_connection_remote_open(event)
      @reconnect.reset if @reconnect
    end

    def on_transport_tail_closed(event)
      self.on_transport_closed(event)
    end

    def on_transport_closed(event)
      if !@connection.nil? && !(@connection.state & Qpid::Proton::Endpoint::LOCAL_ACTIVE).zero?
        if !@reconnect.nil?
          event.transport.unbind
          delay = @reconnect.next
          if delay == 0
            self.connect(@connection)
          else
            event.reactor.schedule(delay, self)
          end
        else
          @connection = nil
        end
      end
    end

    def on_timer_task(event)
      self.connect(@connection)
    end

    def on_connection_remote_close(event)
      @connection = nil
    end

    def connect(connection)
      url = @urls.next
      transport = Qpid::Proton::Transport.new
      @opts.each do |k,v|
        case k
        when :user
          connection.user = v
        when :password
          connection.password = v
        when :heartbeat
          transport.idle_timeout = v.to_i
        when :idle_timeout
          transport.idle_timeout = v.(v*1000).to_i
        when :sasl_enabled
          transport.sasl if v
        when :sasl_allow_insecure_mechs
          transport.sasl.allow_insecure_mechs = v
        when :sasl_allowed_mechs, :sasl_mechanisms
          transport.sasl.allowed_mechs = v
        end
      end

      # TODO aconway 2017-08-11: hostname setting is incorrect, reactor only
      connection.hostname = "#{url.host}:#{url.port}"
      connection.user = url.username if url.username && !url.username.empty?
      connection.password = url.password if url.password && !url.password.empty?

      transport.bind(connection)

      if (url.scheme == "amqps") && @ssl_domain
        @ssl = Qpid::Proton::SSL.new(transport, @ssl_domain)
        @ssl.peer_hostname = url.host
      end
    end
  end
end

File: proton-c/bindings/cpp/src/connector.cpp
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

#include "connector.hpp"

#include "proton/connection.hpp"
#include "proton/transport.hpp"
#include "proton/container.hpp"
#include "proton/reconnect_timer.hpp"
#include "proton/task.hpp"
#include "proton/sasl.hpp"
#include "proton/url.hpp"

#include "container_impl.hpp"
#include "proton_bits.hpp"
#include "proton_event.hpp"

#include <proton/connection.h>
#include <proton/transport.h>

namespace proton {

connector::connector(connection&c, const url& a, const connection_options &opts) :
    connection_(c), address_(a), options_(opts), reconnect_timer_(0), transport_configured_(false)
{}

connector::~connector() { delete reconnect_timer_; }

void connector::apply_options() {
    if (!connection_) return;
    options_.apply(connection_);
}

bool connector::transport_configured() { return transport_configured_; }

void connector::reconnect_timer(const class reconnect_timer &rt) {
    delete reconnect_timer_;
    reconnect_timer_ = new class reconnect_timer(rt);
}

void connector::connect() {
    pn_transport_t *pnt = pn_transport();
    transport t(make_wrapper(pnt));
    if (!address_.user().empty())
        connection_.user(address_.user());
    if (!address_.password().empty())
        connection_.password(address_.password());
    pn_transport_bind(pnt, unwrap(connection_));
    pn_decref(pnt);
    // Apply options to the new transport.
    options_.apply(connection_);
    // if virtual-host not set, use host from address as default
    if (!options_.is_virtual_host_set())
        pn_connection_set_hostname(unwrap(connection_), address_.host().c_str());
    transport_configured_ = true;
}

void connector::on_connection_local_open(proton_event &) {
    connect();
}

void connector::on_connection_remote_open(proton_event &) {
    if (reconnect_timer_) {
        reconnect_timer_->reset();
    }
}

void connector::on_connection_init(proton_event &) {
}

void connector::on_transport_tail_closed(proton_event &e) {
    on_transport_closed(e);
}

void connector::on_transport_closed(proton_event &) {
    if (!connection_) return;
    if (connection_.active()) {
        if (reconnect_timer_) {
            pn_transport_unbind(unwrap(connection_.transport()));
            transport_configured_ = false;
            int delay = reconnect_timer_->next_delay(timestamp::now());
            if (delay >= 0) {
                if (delay == 0) {
                    // log "Disconnected, reconnecting..."
                    connect();
                    return;
                }
                else {
                    // log "Disconnected, reconnecting in " <<  delay << " milliseconds"
                    static_cast<container_impl&>(connection_.container()).schedule(delay, this);
                    return;
                }
            }
        }
    }
    pn_connection_release(unwrap(connection_));
    connection_  = 0;
}

void connector::on_timer_task(proton_event &) {
    connect();
}

}
